File: src/webengine/libqwebenginescript.cpp
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QUrl>
#include <QWebEngineScript>
#include <qwebenginescript.h>
#include "libqwebenginescript.h"
#include "libqwebenginescript.hxx"

QWebEngineScript* QWebEngineScript_new() {
    return new QWebEngineScript();
}

QWebEngineScript* QWebEngineScript_new2(QWebEngineScript* other) {
    return new QWebEngineScript(*other);
}

void QWebEngineScript_OperatorAssign(QWebEngineScript* self, QWebEngineScript* other) {
    self->operator=(*other);
}

libqt_string QWebEngineScript_Name(const QWebEngineScript* self) {
    QString _ret = self->name();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

void QWebEngineScript_SetName(QWebEngineScript* self, libqt_string name) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    self->setName(name_QString);
}

QUrl* QWebEngineScript_SourceUrl(const QWebEngineScript* self) {
    return new QUrl(self->sourceUrl());
}

void QWebEngineScript_SetSourceUrl(QWebEngineScript* self, QUrl* url) {
    self->setSourceUrl(*url);
}

libqt_string QWebEngineScript_SourceCode(const QWebEngineScript* self) {
    QString _ret = self->sourceCode();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

void QWebEngineScript_SetSourceCode(QWebEngineScript* self, libqt_string sourceCode) {
    QString sourceCode_QString = QString::fromUtf8(sourceCode.data, sourceCode.len);
    self->setSourceCode(sourceCode_QString);
}

int QWebEngineScript_InjectionPoint(const QWebEngineScript* self) {
    return static_cast<int>(self->injectionPoint());
}

void QWebEngineScript_SetInjectionPoint(QWebEngineScript* self, int injectionPoint) {
    self->setInjectionPoint(static_cast<QWebEngineScript::InjectionPoint>(injectionPoint));
}

unsigned int QWebEngineScript_WorldId(const QWebEngineScript* self) {
    return static_cast<unsigned int>(self->worldId());
}

void QWebEngineScript_SetWorldId(QWebEngineScript* self, unsigned int worldId) {
    self->setWorldId(static_cast<unsigned int>(worldId));
}

bool QWebEngineScript_RunsOnSubFrames(const QWebEngineScript* self) {
    return self->runsOnSubFrames();
}

void QWebEngineScript_SetRunsOnSubFrames(QWebEngineScript* self, bool on) {
    self->setRunsOnSubFrames(on);
}

bool QWebEngineScript_OperatorEqual(const QWebEngineScript* self, QWebEngineScript* other) {
    return (*self == *other);
}

bool QWebEngineScript_OperatorNotEqual(const QWebEngineScript* self, QWebEngineScript* other) {
    return (*self != *other);
}

void QWebEngineScript_Swap(QWebEngineScript* self, QWebEngineScript* other) {
    self->swap(*other);
}

void QWebEngineScript_Delete(QWebEngineScript* self) {
    delete self;
}

File: src/webengine/libqwebenginescript.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qwebenginescript_enums = enums;
const std = @import("std");

/// https://doc.qt.io/qt-6/qwebenginescript.html
pub const qwebenginescript = struct {
    /// New constructs a new QWebEngineScript object.
    ///
    ///
    pub fn New() QtC.QWebEngineScript {
        return qtc.QWebEngineScript_new();
    }

    /// New2 constructs a new QWebEngineScript object.
    ///
    /// ``` other: QtC.QWebEngineScript ```
    pub fn New2(other: ?*anyopaque) QtC.QWebEngineScript {
        return qtc.QWebEngineScript_new2(@ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#operator-eq)
    ///
    /// ``` self: QtC.QWebEngineScript, other: QtC.QWebEngineScript ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QWebEngineScript_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#name)
    ///
    /// ``` self: QtC.QWebEngineScript, allocator: std.mem.Allocator ```
    pub fn Name(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWebEngineScript_Name(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qwebenginescript.Name: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#setName)
    ///
    /// ``` self: QtC.QWebEngineScript, name: []const u8 ```
    pub fn SetName(self: ?*anyopaque, name: []const u8) void {
        const name_str = qtc.struct_libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        qtc.QWebEngineScript_SetName(@ptrCast(self), name_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#sourceUrl)
    ///
    /// ``` self: QtC.QWebEngineScript ```
    pub fn SourceUrl(self: ?*anyopaque) QtC.QUrl {
        return qtc.QWebEngineScript_SourceUrl(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#setSourceUrl)
    ///
    /// ``` self: QtC.QWebEngineScript, url: QtC.QUrl ```
    pub fn SetSourceUrl(self: ?*anyopaque, url: ?*anyopaque) void {
        qtc.QWebEngineScript_SetSourceUrl(@ptrCast(self), @ptrCast(url));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#sourceCode)
    ///
    /// ``` self: QtC.QWebEngineScript, allocator: std.mem.Allocator ```
    pub fn SourceCode(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWebEngineScript_SourceCode(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qwebenginescript.SourceCode: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#setSourceCode)
    ///
    /// ``` self: QtC.QWebEngineScript, sourceCode: []const u8 ```
    pub fn SetSourceCode(self: ?*anyopaque, sourceCode: []const u8) void {
        const sourceCode_str = qtc.struct_libqt_string{
            .len = sourceCode.len,
            .data = sourceCode.ptr,
        };
        qtc.QWebEngineScript_SetSourceCode(@ptrCast(self), sourceCode_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#injectionPoint)
    ///
    /// ``` self: QtC.QWebEngineScript ```
    pub fn InjectionPoint(self: ?*anyopaque) i64 {
        return qtc.QWebEngineScript_InjectionPoint(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#setInjectionPoint)
    ///
    /// ``` self: QtC.QWebEngineScript, injectionPoint: qwebenginescript_enums.InjectionPoint ```
    pub fn SetInjectionPoint(self: ?*anyopaque, injectionPoint: i64) void {
        qtc.QWebEngineScript_SetInjectionPoint(@ptrCast(self), @intCast(injectionPoint));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#worldId)
    ///
    /// ``` self: QtC.QWebEngineScript ```
    pub fn WorldId(self: ?*anyopaque) u32 {
        return qtc.QWebEngineScript_WorldId(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#setWorldId)
    ///
    /// ``` self: QtC.QWebEngineScript, worldId: u32 ```
    pub fn SetWorldId(self: ?*anyopaque, worldId: u32) void {
        qtc.QWebEngineScript_SetWorldId(@ptrCast(self), @intCast(worldId));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#runsOnSubFrames)
    ///
    /// ``` self: QtC.QWebEngineScript ```
    pub fn RunsOnSubFrames(self: ?*anyopaque) bool {
        return qtc.QWebEngineScript_RunsOnSubFrames(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#setRunsOnSubFrames)
    ///
    /// ``` self: QtC.QWebEngineScript, on: bool ```
    pub fn SetRunsOnSubFrames(self: ?*anyopaque, on: bool) void {
        qtc.QWebEngineScript_SetRunsOnSubFrames(@ptrCast(self), on);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#operator-eq-eq)
    ///
    /// ``` self: QtC.QWebEngineScript, other: QtC.QWebEngineScript ```
    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QWebEngineScript_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#operator-not-eq)
    ///
    /// ``` self: QtC.QWebEngineScript, other: QtC.QWebEngineScript ```
    pub fn OperatorNotEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QWebEngineScript_OperatorNotEqual(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#swap)
    ///
    /// ``` self: QtC.QWebEngineScript, other: QtC.QWebEngineScript ```
    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QWebEngineScript_Swap(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginescript.html#dtor.QWebEngineScript)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QWebEngineScript ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QWebEngineScript_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qwebenginescript.html#types
pub const enums = struct {
    pub const InjectionPoint = enum {
        pub const Deferred: i32 = 0;
        pub const DocumentReady: i32 = 1;
        pub const DocumentCreation: i32 = 2;
    };

    pub const ScriptWorldId = enum {
        pub const MainWorld: i32 = 0;
        pub const ApplicationWorld: i32 = 1;
        pub const UserWorld: i32 = 2;
    };
};
